Remove Goldshell ASIC promo banner and modal (#95)
tari-project/tari-dot-com: _updates/2021-11-18-update-64.md

---
layout: update
tag: Developer Update
date: 2021-11-18
author: SW van Heerden
thumbnail: update-background.jpg
title: Atomic swaps
class: subpage
---

## Tari Atomic swaps

> The release of version 0.21.2 the Tari Console wallet now supports atomic swaps with multiple coins. When Tari released
> [TariScript](https://tlu.tarilabs.com/tari/TariScript_for_dummies) on testnet, this allowed much more advanced payment
> ideas than standard vanilla MimbleWimble such as [One-sided payments](https://www.tari.com/updates/2021-06-03-update-57.html)
> and [Atomic swaps](https://tlu.tarilabs.com/protocols/AtomicSwaps). While almost all of these advanced features were
> possible the moment TariScript launched on testnet, let’s be honest handcrafting transactions is complex. 

Reviewer: ```suggestion
With the release of version 0.21.2, the Tari Console wallet now supports atomic swaps with multiple coins. 
[TariScript](https://tlu.tarilabs.com/tari/TariScript_for_dummies) allows us to implement several advanced payment ideas that are difficult or impossible in vanilla MimbleWimble, including [One-sided payments](https://www.tari.com/updates/2021-06-03-update-57.html)
and [Atomic swaps](https://tlu.tarilabs.com/protocols/AtomicSwaps). 

Most of these advanced features were
possible the moment TariScript launched on testnet, but let’s be honest:  Handcrafting transactions is a complex process. 
```

With this release, we added support via CLI and GRPC commands to initiate, claim and refund atomic swaps with all crypto
coins that support [HTLCs](https://hackernoon.com/what-are-hashed-timelock-contracts-htlcs-application-in-lightning-network-payment-channels-14437eeb9345)

> with a SHA256 pre-image. This means that any crypto coin that can unlock some payment via some SHA256 can do atomic swaps
> with Tari, including Bitcoin, Litecoin, and many more. So what does all this mean? Well, with this new release, all your
> atomic swaps are just a command away. We have written a handy example [guide](https://github.com/tari-project/tari/blob/development/docs/src/btc_atomic_swap.md)
> taking you through step by step to do an atomic swap with Bitcoin, the steps are exactly the same for the rest.

Reviewer: ```suggestion
with a SHA256 pre-image. Any crypto coin that can unlock a payment using a SHA256 pre-image can carry out atomic swaps
with Tari, including Bitcoin, Litecoin, and many more. 

So what does all this mean? Well, with this new release, all your
atomic swaps are just a command away. We have written a handy [guide](https://github.com/tari-project/tari/blob/development/docs/src/btc_atomic_swap.md)
taking you through a BTC-tXTR atomic swap step by step. The process for other coins is entirely analogous.
```

Let us dig a bit deeper into how this all functions and what the process is. We can divide this whole 
process into the following steps:

* The Tari initiation.
* The Other coin initiation.
* The other coin finalization.
* The Tari finalization.
* And if something goes wrong somewhere: Refunds. 

One crucial step that is always hinted at but not always directly mentioned is that the two persons doing the atomic
swaps need to communicate out of band with each other to exchange information.

### Tari initiation

> This step is the first step whereby we start the whole processing by sending a Tari HTLC transaction. This step will be
> where you initiate and broadcast the atomic swap HTLC to the network. Your local wallet will provide you with crucial
> information required to move forward, such as the pre-image, pre-image hash, and the hash of the output created on the
> blockchain. The pre-image needs must be kept secret as this is your key to claiming your coins on the other blockchain,
> and revealing it to the other person will break the security of the atomic swap. The hashed pre-image and the hash of
> the output need to be provided to the other person. 

Reviewer: ```suggestion
The first step consists of broadcasting a Tari HTLC transaction. Your wallet provides the information required to move forward, such as the pre-image, pre-image hash, and the hash of the output created on the
blockchain. The pre-image needs must be kept secret as this is your key to claiming your coins on the other blockchain,
and revealing it to the other person will break the security of the atomic swap. The hashed pre-image and the hash of
the output need to be provided to the counterparty. 
```

> The newly created output will be saved in your local wallet as potential time-locked input. More on that later. 

Reviewer: ```suggestion
The newly created output will be saved in your local wallet as a potential time-locked input. More on that later. 
```

### Other coin initiation

>  Here is where the other person needs to commit to their funds on the other blockchain. This transaction creates an HTLC
>  that has the same pre-image as the Tari one. This can be a complicated step as this is where they will need to make the
>  custom transactions for the other blockchain, but luckily we have tools that will create all of this. 

Reviewer: ```suggestion
 Here is where the other person needs to commit to their funds on the other blockchain. This transaction creates an HTLC
 that has the same pre-image as the Tari one. This can be a complicated step as this is where they will need to craft custom transactions on the other blockchain, but luckily we have tools that will create all of this. 
```

### Other coin finalization

> This is where you claim your coins from the other blockchain. Remember that secret pre-image created in the first step;
> here is where you use it to claim your funds on the other blockchain. Again, this step could be complex, but luckily, we
> have tools to simplify this by following the easy [steps](https://github.com/tari-project/tari/blob/development/docs/src/btc_atomic_swap.md). 

Reviewer: ```suggestion
This is where you claim your coins from the other blockchain. Remember that secret pre-image created in the first step?
Again, this step could be complex, but luckily, we
have tools to simplify this by following the easy [steps](https://github.com/tari-project/tari/blob/development/docs/src/btc_atomic_swap.md). 
```

### Tari finalization

> This is where the other person uses the secret pre-image that you just revealed in claiming your coins on the other
> blockchain. They can use this to claim the Tari you put up for the exchange. This is also just a signal step they have
> to call on their wallet using the pre-image and hash of the output. The wallet will search the blockchain for the output
> and transfer it to a wholly-owned UTXO. 

Reviewer: ```suggestion
Finally, the other person uses the secret pre-image that you just revealed in claiming your coins on the other
blockchain. They can use this to claim the Tari you put up for the exchange. This is also just a signal step they have
to call on their wallet using the pre-image and hash of the output. The wallet will search the blockchain for the output
and transfer it to a wholly-owned UTXO. 
```

### Refunds

> The refunds ensure that everyone can reclaim all their funds if something goes wrong and nothing gets lost. On the Tari
> side, the refunds are handled in one of two ways:

Reviewer: ```suggestion
The refunds ensure that everyone can reclaim all their funds if something goes wrong. On the Tari
side, the refunds are handled in one of two ways:
```
